sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/batch/PartitionScopeThresholds.java

```diff
@@ -0,0 +1,150 @@
+import org.slf4j.LoggerFactory;
+import reactor.util.function.Tuple2;
+
+import java.util.UUID;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static com.azure.cosmos.implementation.guava25.base.Preconditions.checkNotNull;
+
+public class PartitionScopeThresholds<TContext> {
+    private final static Logger logger = LoggerFactory.getLogger(PartitionScopeThresholds.class);
+
+    private final String pkRangeId;
+    private final BulkProcessingOptions<TContext> options;
+    private final AtomicInteger targetMicroBatchSize;
+    private final AtomicLong totalOperationCount;
+    private final AtomicReference<CurrentIntervalThresholds> currentThresholds;
+    private final String identifier = UUID.randomUUID().toString();
+    private final double minRetryRate;
+    private final double maxRetryRate;
+    private final double avgRetryRate;
+
+    public PartitionScopeThresholds(String pkRangeId, BulkProcessingOptions<TContext> options) {
+        checkNotNull(pkRangeId, "expected non-null pkRangeId");
+        checkNotNull(options, "expected non-null options");
+
+        this.pkRangeId = pkRangeId;
+        this.options = options;
+        this.targetMicroBatchSize = new AtomicInteger(options.getMaxMicroBatchSize());
+        this.totalOperationCount = new AtomicLong(0);
+        this.currentThresholds = new AtomicReference<>(new CurrentIntervalThresholds());
+        this.minRetryRate = options.getMinTargetedMicroBatchRetryRate();
+        this.maxRetryRate = options.getMaxTargetedMicroBatchRetryRate();
+        this.avgRetryRate = ((this.maxRetryRate + this.minRetryRate)/2);
+    }
+
+    public String getPartitionKeyRangeId() {
+        return this.pkRangeId;
+    }
+
+    private Pair<Boolean, Boolean> shouldReevaluateThresholds(long totalSnapshot, long currentSnapshot) {
+        if (totalSnapshot < 1_000) {
+
+            return Pair.of(currentSnapshot == 100, false);
+        }
+
+        if (totalSnapshot < 10_000) {
+            return Pair.of(currentSnapshot == 1_000, false);
+        }
+
+        return Pair.of(currentSnapshot % 1_000 == 0, currentSnapshot % 10_000 == 0);
+    }
+
+    private void recordOperation(boolean isRetry) {
+        long totalSnapshot = this.totalOperationCount.incrementAndGet();
+        CurrentIntervalThresholds currentThresholdsSnapshot = this.currentThresholds.get();
+        long currentTotalCountSnapshot = currentThresholdsSnapshot.currentOperationCount.incrementAndGet();
+        long currentRetryCountSnapshot;
+        if (isRetry) {
+            currentRetryCountSnapshot = currentThresholdsSnapshot.currentRetriedOperationCount.incrementAndGet();
+        } else {
+            currentRetryCountSnapshot = currentThresholdsSnapshot.currentRetriedOperationCount.get();
+        }
+
+        double retryRate = (double)currentRetryCountSnapshot / currentTotalCountSnapshot;
```

Comment:
The retryRate is not used in this method, probably can move this logic to reevaluateThresholds

Reply:
It is used in line 87

Reply:
but we pass currentRetryCountSnapshot, currentTotalCountSnapshot to that method?